src/Native/Runtime/GCHelpers.cpp

```diff
@@ -220,3 +220,63 @@ COOP_PINVOKE_HELPER(void, RhGetMemoryInfo, (UInt32* highMemLoadThreshold, UInt64
                                                        lastRecordedMemLoad,
                                                        lastRecordedHeapSize, lastRecordedFragmentation);
 }
+
+EXTERN_C REDHAWK_API Int64 __cdecl RhGetTotalAllocatedBytes(Boolean precise)
```

Comment:
It needs to be `EXTERN_C`. The PInvokeTunnel does not work with `COOP_PINVOKE_HELPER`

Reply:
Also, the argument should be `BOOL`

Reply:
What is the PInvokeTunnel and when is it needed? Most of the GC methods appear to be callable without but some need it. I noticed before that the runtime asserts about the PInvokeTunnel being null e.g., when CppCodeGen performed a GC due to the gen0 size hack breaking.

Reply:
The best explanation i've come across is https://github.com/dotnet/corert/blob/4357f9b8be3f5856ddce285c2534b6f3b08752bd/src/Native/Runtime/thread.cpp#L133-L139 but i'm not sure why it excludes using `COOP_PINVOKE_HELPER`, is it just because register spilling?

Reply:
> What is the PInvokeTunnel and when is it needed?

It is needed when we need to be in cooperative mode with crawl-able stack. It is equivalent of `HELPER_METHOD_FRAME` used in CoreCLR. 

I am actually not sure whether the PInvoke tunel is needed here at all. The SuspendAllThreads may need to be called in preemptive mode for this work well. The way to test this is to start multiple threads: Allocate in a loop on a few of them; and call the precise method in a loop on a few of them. Run this for a while (a few minutes) - you should not see any asserts, crashes or hangs. 

> CppCodeGen performed a GC

The GC does not work in CppCodeGen today.